src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,184 @@
+#include "table_scan.hpp"
+#include "resolve_type.hpp"
+
+namespace opossum {
+
+TableScan::TableScan(const std::shared_ptr<const AbstractOperator>& in, const ColumnID column_id,
+                     const ScanType scan_type, const AllTypeVariant search_value)
+    : AbstractOperator(in), _column_id{column_id}, _scan_type{scan_type}, _search_value{search_value} {
+  DebugAssert(_left_input != nullptr, "TableScan operator requires a left input.");
+  DebugAssert(_right_input == nullptr, "There should not be a right input for the TableScan operator.");
+}
+
+std::shared_ptr<const Table> TableScan::_on_execute() {
+  _result_table = std::make_shared<Table>();
+  // A PosList that is shared between all columns of the result table.
+  _pos_list = std::make_shared<PosList>();
+
+  // Creates all columns of old table in the result table and Initializes the result chunks.
+  const auto input = _left_input_table();
+  const auto column_count = input->column_count();
+
+  for (auto column_id = ColumnID{0}; column_id < column_count; ++column_id) {
+    _result_table->add_column_definition(input->column_name(column_id), input->column_type(column_id),
+                                         input->column_nullable(column_id));
+  }
+
+  // Find the correct data type and scan with this knowledge.
+  resolve_data_type(_left_input_table()->column_type(column_id()), [this](auto type) {
+    using Type = typename decltype(type)::type;
+    _scan_table<Type>();
+  });
+
+  // Add ReferenceSegments using the shared PosList.
+  const auto initial_input_chunk = input->get_chunk(ChunkID{0});
+  const auto initial_result_chunk = _result_table->get_chunk(ChunkID{0});
+  for (auto column_id = ColumnID{0}; column_id < column_count; ++column_id) {
+    const auto segment = initial_input_chunk->get_segment(column_id);
+    const auto segment_as_reference_segment = std::dynamic_pointer_cast<ReferenceSegment>(segment);
+    // Use the referenced table of the reference segment, in case we have one.
+    // Otherwise, use the provided input table as the basis.
+    const auto current_input = segment_as_reference_segment ? segment_as_reference_segment->referenced_table() : input;
```

Comment:
When we encounter tables, where segment types change within the same column (e.g. column 0, chunk 0 -> ReferenceSegement, column 0, chunk 1 -> ValueSegment), this should brake. We would need multiple ReferenceSegments in that case, each referencing their respective Segment in the input table.